crashtracker-ffi/src/collector_windows/api.rs

```diff
@@ -0,0 +1,737 @@
+                result.err().unwrap()
+            )
+            .as_str(),
+        );
+        return E_FAIL;
+    }
+
+    output_debug_string("ddog_crasht_exception_event_callback succeeded");
+    S_OK
+}
+
+pub unsafe fn read_wer_context(process_handle: HANDLE, base_address: usize) -> Result<WerContext> {
+    let buffer = read_memory_raw(process_handle, base_address as u64, size_of::<WerContext>())?;
+
+    if buffer.len() != size_of::<WerContext>() {
+        return Err(anyhow!("Failed to read the full WerContext, wrong size"));
+    }
+
+    // Create a MaybeUninit to hold the WerContext
+    let mut wer_context = MaybeUninit::<WerContext>::uninit();
+
+    // Copy the raw bytes from the Vec<u8> into the MaybeUninit<WerContext>
+    let ptr = wer_context.as_mut_ptr();
+    std::ptr::copy_nonoverlapping(buffer.as_ptr(), ptr as *mut u8, buffer.len());
+
+    // Validate prefix and suffix
+    let raw_ptr = wer_context.as_ptr();
+
+    // We can't call assume_init yet because ptr hasn't been set,
+    // and apparently (*raw_ptr).prefix before calling assume_init is undefined behavior
+    let prefix = read_unaligned(addr_of!((*raw_ptr).prefix));
+    let suffix = read_unaligned(addr_of!((*raw_ptr).suffix));
+
+    if prefix != WER_CONTEXT_PREFIX || suffix != WER_CONTEXT_SUFFIX {
+        return Err(anyhow!("Invalid WER context"));
+    }
+
+    let ptr = read_unaligned(addr_of!((*raw_ptr).ptr));
+    let len = read_unaligned(addr_of!((*raw_ptr).len));
+
+    if ptr.is_null() || len == 0 {
+        return Err(anyhow!("Invalid WER context"));
+    }
+
+    // Read the memory in the target process pointed by ptr
+    let buffer = read_memory_raw(process_handle, ptr as u64, len)?;
+
+    // Copy the buffer into a new Box<[u8]>
+    let boxed_slice = buffer.into_boxed_slice();
+
+    // Leak the Box<[u8]> to get a static reference
+    let static_slice = Box::leak(boxed_slice);
+
+    // Create a new WerContext with the leaked static reference
+    let wer_context = WerContext {
+        prefix: WER_CONTEXT_PREFIX,
+        ptr: static_slice.as_ptr(),
+        len,
+        suffix: WER_CONTEXT_SUFFIX,
+    };
+
+    Ok(wer_context)
+}
+
+// https://github.com/microsoft/win32metadata/issues/1044
+#[repr(align(16))]
+#[derive(Default)]
+struct AlignedContext {
+    ctx: CONTEXT,
+}
+
+unsafe fn walk_thread_stack(
+    process_handle: HANDLE,
+    thread_id: u32,
+    modules: &[ModuleInfo],
+) -> Result<StackTrace> {
+    let mut stacktrace = StackTrace::new_incomplete();
+    let thread_handle = OpenThread(THREAD_ALL_ACCESS, false, thread_id)?;
+    let mut context = AlignedContext::default();
+
+    #[cfg(target_arch = "x86_64")]
+    {
+        context.ctx.ContextFlags = CONTEXT_FULL_AMD64;
+    }
+    #[cfg(target_arch = "x86")]
+    {
+        context.ctx.ContextFlags = CONTEXT_FULL_X86;
+    }
+
+    GetThreadContext(thread_handle, &mut context.ctx)?;
+
+    let mut native_frame = STACKFRAME_EX::default();
+    let machine_type: u32;
```

Comment:
Would it be cleaner to initialize this directly in the `cfg` block, and then we wouldn't need the mut / have an uninitilized variable hanging around?

Reply:
Yeah I did that originally but it doesn't work. If I declare the variable in the block then its scope is limited to the block.